# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: security/manager/ssl/TransportSecurityInfo.cpp
# Commit: 019b333ec455
# Full Hash: 019b333ec4552b0bcaefb27f0cb969815cff9b11
# Author: Sean Feng <sefeng@mozilla.com>
# Date: 2019-10-26 09:45:46
# Description:
#   Bug 1590709 - Fix crash in TransportSecurityInfo::ReadCertList r=keeler
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D50557
# ==============================================================================

diff -r 10e1c85b13b0 -r 019b333ec455 security/manager/ssl/TransportSecurityInfo.cpp
--- a/security/manager/ssl/TransportSecurityInfo.cpp	Fri Oct 25 20:27:46 2019 +0000
+++ b/security/manager/ssl/TransportSecurityInfo.cpp	Fri Oct 25 18:20:53 2019 +0000
@@ -449,7 +449,8 @@
     if (!cert) {
       return NS_ERROR_UNEXPECTED;
     }
-    aCertList.AppendElement(cert);
+    RefPtr<nsIX509Cert> castedCert(cert.get());
+    aCertList.AppendElement(castedCert);
   }
   return NS_OK;
 }
@@ -655,16 +656,6 @@
     }
   }
   // END moved from nsISSLStatus
-  // mIsDelegatedCredential added in bug 1562773
-  if (serVersion.EqualsASCII("2")) {
-    rv = aStream->ReadBoolean(&mIsDelegatedCredential);
-    CHILD_DIAGNOSTIC_ASSERT(NS_SUCCEEDED(rv),
-                            "Deserialization should not fail");
-    if (NS_FAILED(rv)) {
-      return rv;
-    }
-  }
-
   if (!serVersion.EqualsASCII("3")) {
     // The old data structure of certList(nsIX509CertList) presents
     rv = ReadCertList(aStream, mFailedCertChain);
@@ -682,6 +673,16 @@
     NS_ENSURE_SUCCESS(rv, rv);
   }
 
+  // mIsDelegatedCredential added in bug 1562773
+  if (serVersion.EqualsASCII("2") || serVersion.EqualsASCII("3")) {
+    rv = aStream->ReadBoolean(&mIsDelegatedCredential);
+    CHILD_DIAGNOSTIC_ASSERT(NS_SUCCEEDED(rv),
+                            "Deserialization should not fail");
+    if (NS_FAILED(rv)) {
+      return rv;
+    }
+  }
+
   return NS_OK;
 }
 